File: transaction-service/controller/transaction_controller.py
import json
from typing import List

from flask import Blueprint, request
from flask_cors import cross_origin
from werkzeug.exceptions import BadRequest

from service.transaction_service import TransactionService
from utils.datetime_utils import get_datetime_or_none
from datetime import datetime

blueprint = Blueprint('transaction_controller', __name__,url_prefix='/transaction/v1')
service = TransactionService()

@blueprint.route(rule="/healthcheck", methods=['GET'])
def healthcheck():
    return "Transaction API is up and running!"




@blueprint.route(rule="/fetch_all", methods=['GET'])
@cross_origin()
def fetch_all_transactions_by_user_ids_start_date_and_end_date() -> str:
    user_account_ids: List[int] = validate_and_get_user_account_ids()
    start_date: str = validate_and_get_request_date(request_key='start_date')
    end_date: str = validate_and_get_request_date(request_key='end_date')
    print(f"Received request to fetch transactions for user ids {user_account_ids}, start date {start_date} and end date {end_date}")
    result = [record.__dict__ for record in service.fetch_all(user_account_ids=user_account_ids, start_date=start_date, end_date=end_date)]
    return json.dumps(result)


def validate_and_get_user_account_ids() -> List[int]:
    if 'user_account_ids' not in request.args.keys():
        raise BadRequest(f"Please pass user_account_ids in request")
    user_account_ids: List[int] = []
    user_account_id_list: List[str] = request.args.getlist('user_account_ids')
    try:
        user_account_ids = [int(i) for i in user_account_id_list]
    except Exception as e:
        raise BadRequest(
            f"Invalid user_account_ids passed in request - {user_account_id_list}. Please pass valid integer ids")

    if len(user_account_ids) == 0:
        raise BadRequest(f"Please pass valid list of user_account_ids in request")
    return user_account_ids

def validate_and_get_request_date(request_key):

    if request_key not in request.args.keys():
        raise BadRequest(f"Please pass key - {request_key} in request")
    date_str: str = request.args.get(request_key)
    date: datetime =  get_datetime_or_none(datetime_string=date_str, format='%Y-%m-%d')
    if date is None:
        raise BadRequest(f"Invalid date passed in request - {date_str}. Valid date format - yyyy-MM-dd")
    return date_str

File: transaction-service/src/main/java/com/finance/manager/transaction_service/controller/TransactionController.java

package com.finance.manager.transaction_service.controller;

import com.finance.manager.transaction_service.dto.Transaction;
import com.finance.manager.transaction_service.dto.TransactionsDto;
import com.finance.manager.transaction_service.service.TransactionService;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.http.HttpStatusCode;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.server.ResponseStatusException;

import java.time.LocalDate;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

@RestController
@RequestMapping("/transaction")
@CrossOrigin
@RequiredArgsConstructor
public class TransactionController {

    private final TransactionService transactionService;
    private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd");
    private static final Logger logger = LoggerFactory.getLogger(TransactionService.class);


    @PostMapping("/v1/save_all")
    public void saveAll(@RequestBody List<Transaction> transactions) {
        transactionService.saveAll(transactions);
    }

    @GetMapping("/v1/healthcheck")
    public String healthcheck() {
        return "Transaction API is up and running!";
    }

    @GetMapping("/v1/fetch_all")
    public ResponseEntity<TransactionsDto> fetchAllTransactionsByUserIdsStartDateAndEndDate(
            @RequestParam("user_account_ids") List<String> userAccountIdsList,
            @RequestParam("start_date") String startDate,
            @RequestParam("end_date") String endDate,
            @RequestParam(value = "category_ids", required = false) List<String> categoryIdsList,
            @RequestParam(value = "debit_credit_indicator", required = false) String debitCreditIndicator
    ) {

        List<Integer> userAccountIds = validateAndGetUserAccountIds(userAccountIdsList);
        validateAndGetRequestDate(startDate);
        validateAndGetRequestDate(endDate);

        // Parse optional category IDs
        Set<Integer> categoryIds = null;
        if (categoryIdsList != null && !categoryIdsList.isEmpty()) {
            categoryIds = new HashSet<>();
            try {
                for (String categoryId : categoryIdsList) {
                    categoryIds.add(Integer.parseInt(categoryId));
                }
            } catch (NumberFormatException e) {
                throw new ResponseStatusException(HttpStatusCode.valueOf(400),
                        "Invalid category_ids passed in request - " + categoryIdsList + ". Please pass valid integer ids");
            }
        }

        logger.info("Received request to fetch transactions for user ids {}, start date {}, end date {}, debit_credit_indicator = {} {}",
                userAccountIds, startDate, endDate, debitCreditIndicator,
                categoryIds != null ? ", filtered by categories: " + categoryIds : "");

        TransactionsDto transactions = transactionService.getUserTransactions(userAccountIds, startDate, endDate, categoryIds, debitCreditIndicator);
        return ResponseEntity.ok(transactions);
    }

    private List<Integer> validateAndGetUserAccountIds(List<String> userAccountIdsList) {
        if (userAccountIdsList == null || userAccountIdsList.isEmpty()) {
            throw new ResponseStatusException(HttpStatusCode.valueOf(400), "Please pass valid list of user_account_ids in request");
        }

        try {
            List<Integer> userAccountIds = userAccountIdsList.stream()
                    .map(Integer::parseInt)
                    .toList();

            if (userAccountIds.isEmpty()) {
                throw new ResponseStatusException(HttpStatusCode.valueOf(400), "Please pass valid list of user_account_ids in request");
            }

            return userAccountIds;
        } catch (NumberFormatException e) {
            throw new ResponseStatusException(HttpStatusCode.valueOf(400),
                    "Invalid user_account_ids passed in request - " + userAccountIdsList + ". Please pass valid integer ids");
        }
    }

    private void validateAndGetRequestDate(String dateStr) {
        if (dateStr == null || dateStr.isEmpty()) {
            throw new ResponseStatusException(HttpStatusCode.valueOf(400), "Please pass valid date in request");
        }

        try {
            // Parse the date to validate format
            LocalDate.parse(dateStr, DATE_FORMATTER);
        } catch (DateTimeParseException e) {
            throw new ResponseStatusException(HttpStatusCode.valueOf(400),
                    "Invalid date passed in request - " + dateStr + ". Valid date format - yyyy-MM-dd");
        }
    }
}

